verification.py

```diff
@@ -1024,6 +1075,348 @@ def f():
+                rate = random.uniform(0.1, 0.5)
+                index = tuple(np.random.choice(
+                        range(num_populations), size=2, replace=False))
+                demographic_events.append(
+                        msprime.MigrationRateChange(
+                                time=t, rate=rate, matrix_index=index))
+
+        ## Collect all pops together to control coalescence times for DTWF
+        for i in range(1, N):
+            demographic_events.append(
+                    msprime.MassMigration(
+                        time=t_max, source=i, destination=0, proportion=1.0))
+
+        demographic_events.append(
+                msprime.PopulationParametersChange(
+                        time=t_max, initial_size=1000,
+                        growth_rate=0.01, population_id=0))
+
+        def f():
+            self.run_dtwf_coalescent_comparison(
+                key,
+                migration_matrix=migration_matrix,
+                population_configurations=population_configurations,
+                demographic_events=demographic_events,
+                num_replicates=num_replicates,
+                recombination_map=recombination_map)
+        self._instances[key] = f
+
+
+    def run_dtwf_slim_comparison(self, test_name, slim_args, **kwargs):
+
+        df = pd.DataFrame()
+
+        kwargs["model"] = "dtwf"
+        print("Running: ", kwargs)
+        replicates = msprime.simulate(**kwargs)
+        data = collections.defaultdict(list)
+        for ts in replicates:
+            t_mrca = np.zeros(ts.num_trees)
+            for tree in ts.trees():
+                t_mrca[tree.index] = tree.time(tree.root)
+            data["tmrca_mean"].append(np.mean(t_mrca))
+            data["num_trees"].append(ts.num_trees)
+            data["model"].append("dtwf")
+
+        basedir = os.path.join("tmp__NOBACKUP__", test_name)
+        if not os.path.exists(basedir):
+            os.mkdir(basedir)
+
+        slim_script = os.path.join(basedir, "slim_script.txt")
+        outfile=os.path.join(basedir, "slim.trees")
+        slim_args['OUTFILE'] = outfile
+        write_slim_script(slim_script, slim_args)
+
+        cmd = "slim " + slim_script
+        i = 0
+        while i < kwargs['num_replicates']:
+        # for i in range(kwargs['num_replicates']):
+            if i % 10 == 0:
+                print(i)
+            output = subprocess.check_output(cmd, shell=True)
+            ts = msprime.load(outfile)
+            ts = subsample_simplify_slim_treesequence(ts, slim_args['sample_sizes'])
+
+            t_mrca = np.zeros(ts.num_trees)
+            try:
+                for tree in ts.trees():
+                    t_mrca[tree.index] = tree.time(tree.root) + slim_args['NGENS']
+            except ValueError:
+                ## Assuming this is due to multiple roots in the tree - 
+                ## increase number of generations and run again
+                slim_args['NGENS'] = slim_args['NGENS'] * 2
```

Comment:
This has been tricky. Implementing coalescent checking broke the validation by some means I haven't found yet... both tmrca and num_trees were distorted.

I'm a bit hesitant to put too much work into understand this at the moment since SLiM tree sequence recording seems to be under active development. At least versions 3.0 and 3.1 required different handling of node times to give a proper comparison to DTWF - in 3.0 times were (coalescence_time - num_gens_simulated), which would always be negative.

What do you think? It's definitely slower than it needs to be, but functional